### DISCUSSION TITLE
Add ability to synchronize empty directories

### DISCUSSION THREAD

**Discussion by gmk57** [doublecmd/doublecmd#1163] [Suggestions]
For a long time TC had an option to synchronize not only files, but the directories too. For asymmetric sync this means deleting the dir on the right side if it does not exist on the left side.

Please, please add this feature to Double Commander. It would allow to use DC as a full-fledged sync and even semi-automatic backup tool. Currently, a lot of obsolete empty dirs accumulate in the destination over time.

There are [multiple](https://doublecmd.h1n.ru/viewtopic.php?f=7&t=7000) [suggestions](https://doublecmd.h1n.ru/viewtopic.php?f=7&t=6795) [about this](https://doublecmd.h1n.ru/viewtopic.php?f=7&t=4167) [on the forum](https://doublecmd.h1n.ru/viewtopic.php?f=7&t=7054), but no registered issue. So here it is. :)

BTW, many thanks for your great program! Without it, I probably wouldn't be able to switch to Linux.

**Comment by ags1234** [comment]
Excellent request! I want this, too!

**Comment by ags1234** [comment]
Any news on this bug?
Manifestation is like this: I have these folders on the right and synchronize function does not detect them. Using the latest build.
![Screenshot - 21 04 2022 , 17_54_21](https://user-images.githubusercontent.com/21011513/164486809-60bef23c-a376-45ae-88f2-29c3f0fc250e.png)

**Comment by alexx2000** [comment]
No. It's hard to add it.

**Reply by qlahcim** [reply]
@alexx2000  Could you elaborate your strict negative answer in more details?

**Comment by ltguillaume** [comment]
I'd say this is a critical part of synchronizing directories. Moreover, I would say most users will expect this behavior from a feature that is literally named "Synchronize directories", so the current lack of implementation could lead to a lot of confusion...

**Comment by qlahcim** [comment]
I would like to ask for this "empty directories" feature, too. I just solve terrible situation in which a lot (1000's) of empty directories left after Sync.

This feature is really very important!!! 

BTW: Is there any file commander for linux with this feature implemented???

**Comment by gmk57** [comment]
IIRC, [FreeFileSync](https://freefilesync.org/) had this feature, but I haven't used it for ages. :man_shrugging: 

As a temporary workaround, after syncing with Double Commander, I run
```
rsync -rtv --delete --omit-dir-times SOURCE_DIR TARGET_DIR
```
(some of the flags are needed because target is an external HDD formatted in NTFS)

**Reply by qlahcim** [reply]
Thanks for FreeFileSync hint!!! I think the best solution (most user-friendly) in the current state of doublecmd is to use FreeFileSync as external Sync tool, see: https://doublecmd.h1n.ru/viewtopic.php?p=12883&hilit=freefilesync#p12883

But, in general, the **lack of fully functional Directory-File Sync with empty directories feature is definitely the biggest drawback of DoubleCommander**!!!

**Comment by ags1234** [comment]
> But, in general, the lack of fully functional Directory-File Sync with empty directories feature is definitely the biggest drawback of DoubleCommander!!!

I completely agree to this (https://github.com/doublecmd/doublecmd/discussions/1163#discussioncomment-6583793).
For my use case scenario, I would also add: the inability to move files to zip and rar archives, in spite of having the "Move to archive" checkbox in the "Pack files" dialog (https://github.com/doublecmd/doublecmd/issues/601#).

Because of these two I cannot completely switch to Double Commander, and I use Total Commander (shareware, 1 developer, insanely optimized in all areas, written in Delphi for 32 bit and FPC/Lazarus for 64 bit IIRC, is superb and does everything) and Unreal Commander (donationware, 1 developer, written in Delphi, fast, handles queues very well, has good synchronize directories function, but does not move files to archives).

**Comment by qlahcim** [comment]
Does the developer's quiet here mean, that in the near (or may be far future) is not possible to expect any progress on this topic???

**Comment by alexx2000** [comment]
I started to implement it sometime ago. But I can’t say when the work will be completed and is it ever will be completed at all.

**Reply by qlahcim** [reply]
@alexx2000  just want to say, that this feature is really crucial and should be solved with high priority. The proper directory Sync functionality is an essential part of work for any (at least a little advanced) user. Please, try to reconsider your development plans, because DoubleCommander is really excellent tool and proper Syncing move this file manager on the top!!!

**Comment by Skif-off** [comment]
> I started to implement it sometime ago.

Will it look like in TC? I found the description of "Empty directories" in the TC documentation:

> Also synchronize empty directories, not just files:
> 1. In normal (symmetric) mode, directories missing on one side will be copied from the other side.
> 2. In asymmetric mode, directories missing on the left side will be deleted on the right side, and directories missing on the right will be copied from the left side.

**Reply by alexx2000** [reply]
Yes.

**Comment by qlahcim** [comment]
@alexx2000 Any near future plans regarding this topic?

**Reply by alexx2000** [reply]
No specific deadlines.

**Comment by qlahcim** [comment]
@alexx2000 Still no progress on this topic? Why is this crucial functionality still ignored?

**Reply by antodc** [reply]
The developer has done an amazing work with this app. For free. Maybe you could be more polite. Even grateful for what you got until now, if you understand the meaning of that word.

**Reply by ltguillaume** [reply]
I don't think it was as impolite as you make it out to be. That said, while I myself do appreciate the complexity of a reliable synchronization process, I am also eagerly awaiting the implementation of this.

**Reply by qlahcim** [reply]
@antodc Did you really read my latest comment with elementary willingness to understand it, or just you have a bad mood? If you understand the meaning of that word? :)

@ltguillaume Thank you for rewording my comment to make it more understandable.

**Reply by ags1234** [reply]
Of course we all appreciate the immense work put into DC and the results of that work.

There are two problems with the synchronize function not supporting empty directories.

One is obvious: it is missing functionality. After the synchronization says it finished successfully, the users are left with two structures that they believe to be identical but are in fact different.

The other is misleading: because in Synchronize directories window there is no warning informing the user about the fact that this function is incomplete. Anyone, being technical or not, would correctly assume it is complete and the comparison is happening as expected.
It's a hidden pit in which the users fall. It happened to me, I had to use Total Commander after I used DC, to get to the results I expected.

For more advanced users having the ability to reliably compare and synchronize nested structures is very important. Having the file manager you like and use do that is very convenient.

To me @michalkvasnicka question is completely understandable.

**Reply by alexx2000** [reply]
> @alexx2000 Still no progress on this topic? Why is this crucial functionality still ignored?

Not yet. Because it is a very boring and uninteresting task (for me). That's why I keep putting it off.

**Reply by qlahcim** [reply]
@alexx2000  Good to know ...

**Reply by antodc** [reply]
> @antodc Did you really read my latest comment with elementary willingness to understand it, or just you have a bad mood? If you understand the meaning of that word? :)

@michalkvasnicka. Just to make it clear. I perfectly understood your comment and even agree with the relevance of this issue and share your interest for a solution. However, in my opinion the tone you used is totally inappropriate. I think that words like 'please', 'kindly', 'may', 'thank you' - none of which you used - are strongly needed when asking other people to spend their time to solve our problems.

**Reply by qlahcim** [reply]
To be perfectly clear:
I think the whole problem is now closed once and for all by @alexx2000's last statement: **"_Not yet. Because it is a very boring and uninteresting task (for me). That's why I keep putting it off_."**, regardless of the fact that it is clearly a rather serious issue that other file commanders (e.g. Krusader, TC, ...) solve completely and elegantly. So, any doubts how it is [hard](https://github.com/doublecmd/doublecmd/discussions/1163#discussioncomment-6583794) are irrelevant.

And finally, it has absolutely nothing to do with the level of "decency" users use in their issue resolution requests. Users just have to accept this fact ... :(

**Reply by ags1234** [reply]
Just to make it ultra-mega-giga super-duper clear: @alexx2000 said „Not yet”. There is still hope :))

Maybe someone around here that knows Pascal can help start things up?

**Comment by antodc** [comment]
To keep this useful (for me) discussion going, I used for years Kdiff3 to compare/syncronize directories. Mostly on Windows (https://kdiff3.sourceforge.net for an old version that I still use), but also on Linux. The most recent versions, also for MacOS are available here https://apps.kde.org/it/kdiff3.
Thanks to one comment above by @qlahcim about FreeFileSync, I was able to create a button to start Kdiff3 from DC. Very easy indeed.
Icon -> 'path to kdiff3.exe',0
Command -> 'path to kdiff3.exe'
Parameters -> %Dl %Dr
In that way it compares the content of the left panel with the content of the right panel. Several options for the comparison are available. And you can also syncronize the panels, including empty directories.
In my view, a satisfactory solution while waiting for something better in DC.

**Reply by ags1234** [reply]
Downloaded kdiff3-master-1754-windows-cl-msvc2022-x86_64 Windows build from here: https://cdn.kde.org/ci-builds/sdk/kdiff3/master/windows/
KDiff3 looks interesting as an alternative to DC's incomplete dir sync. Will test it later.
Right from the start I see some disadvantages for me:
- external tool, though probably a very good one
- adds 411 MiB to the 46MiB of DC  :-/
- not portable, as DC is.

Also, looking at the link @antodc gave above, I noticed a side recommendation from sourceforge: WinMerge. Should check this as well.

**Reply by ags1234** [reply]
And a few commercial alternatives that I know of for comparing files or directory structures, maybe someone is interested:
- ExamDiff 
- Beyond Compare
- UltraCompare

**Reply by antodc** [reply]
>     * adds 411 MiB to the 46MiB of DC  :-/

Maybe I should have mentioned. The main reason why I keep using the older 0.9.98 version is its small(er) size. Just about 25Mb in my case.

**Reply by ltguillaume** [reply]
For Windows users (@antodc @ags1234), don't forget WinMerge:
- Open source
- Compares directories, archives and files
- Is only about 18MB big
- _Is_ portable (use a `winmerge.ini` file)

For Linux user, don't forget Wine + WinMerge 😉

**Reply by ags1234** [reply]
Lieutenant guillaume, :))
I tested WinMerge and I like it. I don't know which of the two (KDiff and WinMerge) has better optimizations in regards to compare algorithms and speed in very large structures, but WinMerge seems to be a good choice for now (portable, small size).

Also, I guess you've been lieutenant for a while now. For your contribution to the greater good of mankind, I propose we raise your military rank to captain :))
All in favor?

**Reply by qlahcim** [reply]
Kdiff3 is not suitable solution for non KDE Linux distros. Kdiff3 using a lot of Qt libs, which must be installed too, on any non KDE (e.g. GNOME) based Linux distro. In that case is a better way to use directly [Krusader](https://krusader.org/) file commander, which has very good internal [Synchronizer](https://docs.kde.org/stable5/en/krusader/krusader/synchronizer.html) tool, but there is the same problem with a bunch of Qt libs dependency.

**Reply by qlahcim** [reply]
My latest progress on this topic:

I just found a viable solution which works well, at least for me .... :)

[Meld](http://meldmerge.org/) looks like a pretty powerful, simple, small and effective tool (fully multi-platform: Win, MacOS, Linux):

Tooltip: DirCompare (Meld)
Command: meld
Parameters: %Dl %Dr

Works well with empty directories, too!!!!

Tooltip: FileCompare (Meld)
Command: meld
Parameters: %pl %pr

**Reply by ltguillaume** [reply]
>When Meld shows differences between files, it shows both files as they would appear in a normal text editor. It does not insert additional lines so that the left and right sides of a particular change are the same size. There is no option to do this. 

This is pretty annoying imho

Also, it's not possible to make it portable AFAIK, and its got a pretty basic feature set when compared to WinMerge, while being about 4 times the size of WinMerge.

I also use the image comparison of WinMerge oftentimes, which visually compares two (or 3?) images, quite useful.

**Reply by ags1234** [reply]
Completely off-topic:
@ltguillaume, I sometimes have to do some image comparison, including comparing **similar** images. These help with the cleaning:
- AllDup_4.5.54_Portable
- Czkawka_7.0.0
- Duplicate.Cleaner.Pro.5.22.0.Portable
- Find.Same.Images.OK-4.55

**Comment by qlahcim** [comment]
Is there anybody who has some successful experience with external FreeFileSync call from DC on Linux?

Setup:
Icon : /opt/FreeFileSync/Resources/FreeFileSync.png
Tooltip : FreeFileSync
Command : FreeFileSync
Parameters : -leftdir %Dl -rightdir %Dr

does not work with latest version of FreeFileSync or DC on Linux

**Confirmed by FFS developer**: FFS does not support standard command-line parameters, so there is no chance to use it via DC Toolbar!!! 

================= **Final update:** =================
Parameters : -DirPair %Dl  %Dr

This works!!! So, directory sync (but nothing else) could be performed from DC by FreeFileSync very fast and reliably. But everything else is terribly complicated, because FFS use batch and global XML config files.

But, frankly speaking, internal "directory sync" at DC with proper behavior (empty folders issue) could be the best solution at all ... :)

**Reply by j2969719** [reply]
>Command : FreeFileSync

you sure about this one? i mean the uppercase doesnt give the impression that typing this in the terminal will launch something. try absolute path to the executable instead

**Comment by RiffSkill** [comment]
I support the request! The function of deleting directories during synchronization is very necessary! Without it, the function "Synchronization" does not correspond to its name!

**Comment by ags1234** [comment]
Any news on implementing the much needed search for empty directories?
If @alexx2000 does not work on this, is anyone else willing to take a look into it?

**Comment by rich2014** [comment]
`empty directories` has been added.

try DC [1.3.0 (13360+)](https://github.com/doublecmd/snapshots/releases/latest) please

**Reply by qlahcim** [reply]
Good news ... Thanks @rich2014 !!! 

@alexx2000 When will be this 5 years eagerly awaited DC version fully tested and available for standard download (upgrade)?

**Reply by j2969719** [reply]
depends on how many of those who have been waiting for five years volunteer to fully test [snapshots](https://github.com/doublecmd/snapshots/releases) with this feature and provide feedback to the developers
<img src="https://github.com/user-attachments/assets/42cb3d86-4bb7-461c-913f-e93d3df5c918" />

**Reply by qlahcim** [reply]
OMG... And I stupidly thought that the developers would test it themselves when someone tried to solve it for them?!
<img width="644" height="644" alt="image" src="https://github.com/user-attachments/assets/817cfbb7-4f42-4b0f-ba12-59608599dc88" />

**Reply by j2969719** [reply]
are you implying that **rich2014** is not a developer or that he should collect feedback for himself despite his request to try a snapshot?

**Reply by qlahcim** [reply]
I meant that the other members of the DC development team should test it first!

So I just asked how long it would take to test the new snapshot and was told that volunteers would have to be found first to provide feedback. My personal experience with regular user testing, which is obviously called for here, or rather providing feedback to developers, is often ineffective.

So I'd rather not ask anything anymore and just wait another 5 years for a version of DC that will finally be able to synchronize even empty directories.

**Reply by rich2014** [reply]
a lot of code has been modified for the feature, so it is essential that everyone test it.

i use the latest 1.3.0 daily.

and, the new version 1.3.0 (13365) has just been released.

**Comment by rich2014** [comment]
DC [1.3.0 (13365)](https://github.com/doublecmd/snapshots/releases/latest) has just been released.

fixed several minor display issues in `Synchronize directories`.

test it please.

**Reply by ags1234** [reply]
@rich2014, this is the most wonderful news! Thank you very much for this important work! 🥇 
I tested the function and it works as intended.  I tested build 13360, not 13365.
One thing I noticed is that the new checkbox state is not saved, so after checking it, synchronizing and closing the window, the checkbox has to be checked again next time the synchronize window is open.
Is it possible to make it so that the last state is saved?

**Reply by rich2014** [reply]
@ags1234  hi, it has been done in 13365.

> One thing I noticed is that the new checkbox state is not saved, so after checking it, synchronizing and closing the window, the checkbox has to be checked again next time the synchronize window is open.
Is it possible to make it so that the last state is saved?

**Reply by j2969719** [reply]
just in case its not intentional: subdirs expand their parents
<img width="1366" height="718" alt="no_empty" src="https://github.com/user-attachments/assets/35b86c55-4f43-4dcc-8e78-c913f4ebb9bc" />
<img width="1366" height="718" alt="empty" src="https://github.com/user-attachments/assets/9c00bff3-4048-43fe-b75d-71d83604651f" />

btw, the last column header is not resizeble, but afaik its an old story like with the lack of a horizontal scrollbar

**Reply by j2969719** [reply]
sorting empty size/date columns produces random results
<img width="1366" height="718" alt="sort" src="https://github.com/user-attachments/assets/6545f762-b6e8-43b7-b1be-a57954c15925" />

**Reply by rich2014** [reply]
if I understand these two images correctly, this was intentional.

you can compare them with TC.

> just in case its not intentional: subdirs expand their parents

**Reply by ags1234** [reply]
### I have the following scenario
- in the right side I have dir1 ... dir 3, all empty dirs
- I do not want them anymore, so I want to use the Delete from the context menu to delete them from the right
### Intended behavior
The dirs get deleted and disappear from the list.
### Observed behavior
Delete function does not do anything to the dirs.

<img width="736" height="401" alt="2026-08-21_183011" src="https://github.com/user-attachments/assets/26c3f100-defb-4feb-b979-053035313277" />

**Reply by rich2014** [reply]
if I understand correctly, you used the Size column on the left to sort, but your example should use the Size column on the right.

> sorting empty size/date columns produces random results

**Reply by j2969719** [reply]
yep, its just that if you intentionally click on the empty left columns, the files on the right are rearranged randomly

**Reply by rich2014** [reply]
@ags1234 

add the feature in 1.3.0 (13367), test it please.

> Delete function does not do anything to the dirs.

**Reply by rich2014** [reply]
as i understand it, isn't this normal?

> yep, its just that if you intentionally click on the empty left columns, the files on the right are rearranged randomly

**Reply by ags1234** [reply]
@rich2014 
Tested 13367, delete function works. Thank you! Thank you! Thank you!

**Comment by rich2014** [comment]
according to the previous design, even if the width of the last column could be adjusted, it would be useless.

regardless of the width of the last column, the filename would always use the remaining space.

> btw, the last column header is not resizeble, but afaik its an old story like with the lack of a horizontal scrollbar

**Comment by ags1234** [comment]
Any reason why the options are not written with a capital letter, like "Subdirs"?
<img width="159" height="109" alt="2026-08-22_201611" src="https://github.com/user-attachments/assets/8382c2b8-dcb7-4e8e-a7a2-cee904d45c61" />

**Reply by ltguillaume** [reply]
And don't have hotkeys assigned, like this:

<img width="459" height="198" alt="image" src="https://github.com/user-attachments/assets/e79a072e-e20b-4802-8d8e-ebd328ab0000" />

(I did this by using a custom language file)

**Reply by rich2014** [reply]
1. it is purely based on the situation of several nearby options. in fact, except for `Subdirs`, all the others start with a lowercase. and none of them contain acceleration.

2. it is probably due to historical reasons. modifying it now would be simple in itself, but the problem is that it would affect translations in all languages.

3. also i feel that hints are missing.

@alexx2000 could you take a look at this?

> Any reason why the options are not written with a capital letter, like "Subdirs"?

> And don't have hotkeys assigned, like this:

**Comment by Skif-off** [comment]
(I don't like screenshots unless they're really necessary, but I don't think I can describe this in words...)
![1.png](https://github.com/user-attachments/assets/8f67aeca-accf-4f95-8d25-56d0f40f6261)

So, six "empty directories", but in reality there's only one "empty folder".
I guess those five "empty directories" should not be in the list.

**Reply by rich2014** [reply]
yes, it maintains consistency with TC

**Comment by Skif-off** [comment]
Hm... I checked TC and see a similar picture.

**Comment by j2969719** [comment]
it might be convenient if ctrl+a / mousedrag also select dir rows with <=> column

**Reply by rich2014** [reply]
@j2969719 

supported, try the latest 1.3.0 please.

**Reply by j2969719** [reply]
kudos! 

some observation: rightclicking a dir now highlights its childs. iirc syncdirs had a quirk regarding selection: if multiple files were selected and you rightclicked somewhere outside the selection, the selection would extend to that file. now when rightclicking  on a dir, the selection extends only to the directory itself excluding its childs

**Reply by rich2014** [reply]
fundamentally speaking, do you think the feature for extending the selection via right-click should be kept?

it doesn't align with standard Grid usage, nor does TC have this functionality.

while it might be useful in certain situations, it is uncommon and could potentially cause confusion.

**Reply by rich2014** [reply]
@alexx2000 

Alex added it at 2da8f38 in 2018, `FIX: Synchronize directories - don't change selection on right mouse click (Qt4, Qt5)`

the title of the commit specifies Qt4 and Qt5, but the code is not limited to Qt.

**Reply by j2969719** [reply]
im just a random guy, and my opinion doesnt carry much weight

if a user relies solely on the mouse and just wanting to launch the differ or viewer this behavior might not be ideal, dont recall any negative feedback about it either though

**Comment by Skif-off** [comment]
It seems to me that a separate background color for directory rows looked more convenient...

**Reply by ags1234** [reply]
I agree. As an example, TagScanner has this and looks good and tidy: https://www.xdlab.ru/en/screen.htm

**Reply by rich2014** [reply]
done!

try 13464+ please.

and Dir selected color can be set in `Options` - `Colors` - `Synchronize Directories`

**Reply by Skif-off** [reply]
It’s something confusing and unnecessary... I meant a solid background color for subdirectories, like the SyncDirs tool used before.

**Reply by rich2014** [reply]
in what way does it conflict with 13464?

could you be more specific?

**Reply by Skif-off** [reply]
Before the synchronization tool was modified, subdirectory names were highlighted with a background color.
Later, the subdirectory names began to look indistinct.
Now it’s not a conflict, they just look bad.

**Reply by rich2014** [reply]
are you referring to 13464?

in 13464, you can set the background color for selected directories and files separately; of course, you can also set them to the same color.

there's nothing else different, is there?

**Reply by ags1234** [reply]
> done!
> 
> try 13464+ please.
> 
> and Dir selected color can be set in `Options` - `Colors` - `Synchronize Directories`

Nice, but the colors can be seen only if the user selects the items in the result of the comparison table. Was this intended?

**Reply by rich2014** [reply]
@ags1234 

it's the `Selected` background color for `Directories`, compared to `Selection` for `Files`.

**Reply by Skif-off** [reply]
@rich2014,

> in 13464, you can set the background color for selected directories and files separately; of course, you can also set them to the same color.
>
> there's nothing else different, is there?


True, but that’s not what I was talking about.
Launch version 1.2.8 and compare the background of the rows with subdirectory paths: it used to be more noticeable and clearer where the subdirectories and where the files were.

**Reply by rich2014** [reply]
> Launch version 1.2.8 and compare the background of the rows with subdirectory paths: it used to be more noticeable and clearer where the subdirectories and where the files were.

i'm a bit confused, do you mean white background color for the directories in 1.2.8?

or perhaps you could provide a screenshot?

**Reply by Skif-off** [reply]
@rich2014

> or perhaps you could provide a screenshot?

For example, [here](https://github.com/doublecmd/doublecmd/discussions/1163#discussioncomment-18145110): do you see that the background color for files and the background color for directories are different?

(I just felt that for working with large list this difference looks better, because more noticeable and clearer than one background color for all rows.)

**Reply by rich2014** [reply]
@Skif-off 

in 1.2.8, the directory actually has no background color—regardless of whether it is selected or not—so it is impossible to tell if it is selected. your screenshot illustrates it.

in 1.3.0  (13464+), the selected directory should has background color to highlight it. shown as following. 

isn't it like that on your end? i recall you use Linux? please provide a screenshot so i can check.

<img width="1854" height="1226" alt="image" src="https://github.com/user-attachments/assets/23723337-b574-4131-8882-47525e15c0a7" />

**Reply by j2969719** [reply]
think theres just a miscommunication. dunno about mac but what skif probably meant is the color in the unselected state (previously judging by the screen it was gray and now its probably white)

**Reply by rich2014** [reply]
@Skif-off @j2969719 

oh, i got it.

the background color for unselected directories varies across different platforms.

fixed in 1.3.0 (13492+).

**Reply by Skif-off** [reply]
Thanks! Both :))

**Comment by Skif-off** [comment]
Strange new changes...
Two directories: [] empty directories > "=" only > Ctrl+A > Delete on both sides > got completely empty directories O_o

(Do not suggest using a mask, this is impossible. Do not suggest modifier keys (Ctrl+, Shift+): more than seven thousand files in approximately 40-50 subdirectories.)

**Reply by rich2014** [reply]
@Skif-off 

the handling of filters will be further refined.

**Reply by Skif-off** [reply]
What do you mean? "All/Directories/Files" switcher, something like the "Show" switchers block?

P.S. It seems to me that the best approach would be to define goals/objectives first, and then make any changes... Right now, the tool gives the impression of being only half a tool and will end up cluttered with workarounds.

P.P.S. I apologize if my opinion seemed harsh.

**Reply by rich2014** [reply]
the use case is simple: aim to stay consistent with the TC unless there is a clear advantage.

`Sync Dirs` is currently working towards this goal step by step; it just takes some time.

**Reply by rich2014** [reply]
@Skif-off 

fixed!

try 13471+ please.

> Strange new changes... Two directories: [] empty directories > "=" only > Ctrl+A > Delete on both sides > got completely empty directories O_o
> 
> (Do not suggest using a mask, this is impossible. Do not suggest modifier keys (Ctrl+, Shift+): more than seven thousand files in approximately 40-50 subdirectories.)

**Reply by Skif-off** [reply]
@rich2014,

> fixed!
>
> try 13471+ please.

Can you try too? I didn't understand what was fixed :) I followed the same steps and I'm seeing the same shit: SyncDirs completely clears out both directories.

I compared the behavior of DC 1.3.0 r13477 and TC 11.57: TC works the same way DC used to.

> step by step; it just takes some time

Why not do it all at once? In fact, SyncDirs is broken, it’s awkward to use, and there’s no way to revert to the way things were, because instead of a single commit/PR, there’s a huge pile of commits...

**Reply by rich2014** [reply]
> Can you try too? I didn't understand what was fixed :) I followed the same steps and I'm seeing the same shit: SyncDirs completely clears out both directories.

of course, I have already tested it. align with TC.
could you provide a test folder?

> Why not do it all at once?

because this is an undiscovered bug—prior to release.

**Reply by Skif-off** [reply]
@rich2014

> could you provide a test folder?

For example, [one](https://github.com/shimmerproject/elementary-xfce/archive/2a57ea3e43e817978b284b40f501057e7d5924ea.zip) and [two](https://github.com/shimmerproject/elementary-xfce/archive/41b209b8e518fcabb86b20b132a80382aeee2fc2.zip).  Unpack > SyncDirs with
```
main:
  [ ] only selected
  [ ] empty directories
  [ ] asymmetric
  [v] Subdirs
  [v] by content
  [v] ignore date
show:
  >          = false
  =          = true
  ≠          = false
  ?          = false
  <          = false
  duplicates = true
  singles    = true
```
Now _Compare_ > Ctrl+A > the right mouse button menu > Delete on both sides.

Try the same with DC 1.2.8.

**Reply by rich2014** [reply]
@Skif-off 

i got it.

there is a bug.

it has been fixed in 1.3.0 (13491+), try it please.

**Reply by Skif-off** [reply]
@rish2014

Still deletes directories, but I didn't intend to delete directories (and don't use the "empty directories" option).

A little oddity: the `elementary-xfce/animations` subdirectory becomes empty, but isn't deleted.

**Reply by rich2014** [reply]
@Skif-off 

are you specifically referring to a particular directory that should not be deleted, or to all directories?

if "empty directories" is not checked, should the directory *not* be deleted—even if it was explicitly selected for deletion and contains no files or subdirectories?

> Still deletes directories, but I didn't intend to delete directories (and don't use the "empty directories" option).

**Reply by rich2014** [reply]
@Skif-off 

since there are no files under `elementary-xfce/animations`, it does not appear in the Grid; therefore, it was not included when you `select all`.

> A little oddity: the elementary-xfce/animations subdirectory becomes empty, but isn't deleted.

**Reply by Skif-off** [reply]
@rich2014

> are you specifically referring to a particular directory that should not be deleted, or to all directories?

I want it to be like before: synchronization works with files, without touching subdirectories (i.e. don't delete the subdirectories).
And if the "empty directories" option is enabled, then empty subdirectories should also be shown in the list (with the same rules as for files, I mean, choice of action, direction of copying and so on).

**Reply by rich2014** [reply]
@Skif-off 

in 13506+, if "empty directories" is not checked, the selected directories will not be deleted when deleting the selected files (even if the user has explicitly selected those directories)

**Comment by antodc** [comment]
First of all, many thanks to all of you that are working to improve the sync tool. The ability to sync empty dirs, in particular, was a very desidered feature.

If anyone has some spare time, I would suggest a few refinements to the UI.

a) Introduce the possibility to auto-adjust the column size by double-clicking on its edge.

b) Center the "< = >" symbols in the name of the central column, given that all elements in that column are centered.

c) The last column on the right ("Name") seems to have a fixed width. Make it adjustable, as is possible for other columns.

d) The item "Subdirs" has a capital letter at the beginning, while all other items start with small letters. (EDIT Funny enough, also in the doc page https://doublecmd.github.io/doc/en/syncdirs.html it is the only item with a cap letter at the beginning).

Finally, one question. Are "duplicates" and "singles" really useful? Maybe I am mistaken, but it seems to me that "duplicates" is equivalent to "=" and "singles" is equivalent to "<" + ">". If this is really the case, the two buttons might be removed. That would lead to a cleaner and less confusing UI, IMHO.

**Reply by alexx2000** [reply]
> Finally, one question. Are "duplicates" and "singles" really useful?

Yes, "duplicates" means files that exists on both sides, "singles" only on one side. This is not equivalent to "=" and "<",">".

**Reply by antodc** [reply]
> > Finally, one question. Are "duplicates" and "singles" really useful?
> 
> Yes, "duplicates" means files that exists on both sides, "singles" only on one side. This is not equivalent to "=" and "<",">".

Fine, I was not correct with "duplicates". Switching "duplicates" on/off is equivalent to switching "=" AND "!=" toghether on/off. If two files are on both sides they are either equal or different. If you like to keep "duplicates", I think that its selection should be reflected in the selection of "=" and "!=", and viceversa. If one sets "duplicates" on, then both "=" and "!=" should get switched on. If "duplicates" is switched off, both "=" and "!=" should get switched off.

For instance, if one selects both "=" and "!=" it is clear (to me) that he wants to see files that are on both sides and are either equal or different. But he actually sees none of them if "duplicates" is not activated. I would suggest that when one selects "=" and "!=", then "duplicates" is automatically turned on.

The current selection of the "Show" options is inconsistent, in my opinion. For "<", "=" and so on there is a "union/or" logic, that is all files that satisfy either of those choices are shown. But for "duplicates" and "singles" there is an "intersection/and" logic. They have to be selected together with other options, otherwise they do not show what they say. If one selects "duplicates" and neither "=" nor "!=", then nothing is shown.

Same reasoning for "singles" and "<"+">".

**Reply by alexx2000** [reply]
> Fine, I was not correct with "duplicates". Switching "duplicates" on/off is equivalent to switching "=" AND "!=" toghether on/off. 
>...
>Same reasoning for "singles" and "<"+">".

No, file can be on the both sides but has "<" or ">" mark (because newer/older). So, these are not equivalent options, but they are interconnected.

**Reply by rich2014** [reply]
@antodc 

it's improved.

this thread is already quite crowded, and your request could easily get lost; please start a separate discussion for your other request.

> b) Center the "< = >" symbols in the name of the central column, given that all elements in that column are centered.

**Comment by eMouses** [comment]
Could you add a setting so that the "Confirm overwrites" checkbox is unchecked by default during synchronization?
<img width="323" height="254" alt="synch" src="https://github.com/user-attachments/assets/de0ff692-dc01-4b4e-a276-aa4514f376dc" />

**Reply by rich2014** [reply]
this thread is already quite crowded, and your request could easily get lost; please start a separate discussion.

**Comment by RiffSkill** [comment]
While the synchronization tool is being finalized, is it possible to immediately add file verification when copying during synchronization, like it is done in TC? There was already a request to add this earlier.

**Reply by rich2014** [reply]
this thread is already quite crowded, your request  will be discussed in #2937
